types/react-i18next/src/context.d.ts

```diff
@@ -24,3 +30,47 @@ export interface i18NextModule {
 }
 
 export const reactI18nextModule: i18NextModule;
+
+export interface ReactI18nContext {
+    i18n: i18n;
+    defaultNS?: string;
+    reportNS?: string;
+    lng: string;
+    t: TranslationFunction;
+}
+
+export const I18nContext: React.Context<ReactI18nContext|undefined>;
+
+// Diff / Omit taken from https://github.com/Microsoft/TypeScript/issues/12215#issuecomment-311923766
+type Omit<T, K extends keyof T> = Pick<T, ({ [P in keyof T]: P } & { [P in K]: never } & { [x: string]: never, [x: number]: never })[keyof T]>;
+
+export type ContextInferableComponentEnhancerWithProps =
+    <P extends { [key: string]: any }>(component: React.ComponentClass<P> | React.StatelessComponent<P>) =>
+        React.ComponentClass<Omit<P, keyof ReactI18nContext> & React.ClassAttributes<never>>;
+
+export function withContext(): ContextInferableComponentEnhancerWithProps;
+
+export interface WithI18nProps {
+    i18n: i18n;
+    t: TranslationFunction;
+    lng: string;
+    i18nOptions: ReactOptions;
+}
+
+export interface WithI18nHocProps {
+    i18nOptions?: ReactOptions;
+}
+
+type ComponentProps<T extends React.ComponentType<any>> =
+    (T extends React.ComponentClass<infer P>
+        ? P
+        : T extends React.SFC<infer P>
+            ? P
+            : {}
+    );
```

Comment:
```ts
type ComponentProps<T extends React.ComponentType<any>> =
  T extends React.ComponentType<infer P> ? P : {}
```

sounds like something that could be useful in `react.d.ts` itself, at least until `infer` is allowed in places other than conditional type conditions.

Reply:
The `any` is required by the way; the type argument is not contravariant so it would not allow a component to receive more props than are enumerated.